Small fix on nnet-cudnn-simple-component.cc; add the user interface a…

tomkocse opened the pull request:
…nd experiment for using cudnn Convolution layer

tomkocse committed:
Small fix on nnet-cudnn-simple-component.cc; add the user interface and experiment for using cudnn Convolution layer

galv commented on src/nnet3/nnet-cudnn-simple-component.cc:
```diff
@@ -232,7 +232,7 @@ void CuDNN3DConvolutionComponent::InitFromConfig(ConfigLine *cfl) {
   if(!cfl->GetValue("pad-z-dim", &pad_z_dim)) {
     pad_z_dim = 0;
   }
-  int32 filter_input_dim = filt_x_dim * filt_y_dim * input_z_dim;
```
Thank you for this fix.

galv commented on src/nnet3/nnet-cudnn-simple-component.cc:
```diff
@@ -638,11 +638,11 @@ void CuDNN3DConvolutionComponent::Write(std::ostream &os, bool binary) const {
   KALDI_ASSERT(filter_dims[1] == input_num_filters_);
   KALDI_ASSERT(filter_dims[0] == num_filters_);
   WriteToken(os, binary, "<FilterXDim>");
-  WriteBasicType(os, binary, filter_dims[2]);
```
I need  to verify whether this is correct. I'm afraid it may be, which means there are other parts of the code which may have the same problem.

galv replied:
This change is correct. Thank you. CUDNN's 3D convolution assumes tensors are laid out as "NCDHW" format, with the first letter corresponding to index 0, and the last corresponding to index 4.

I will look through the rest of the code to verify that I correctly follow CUDNN's indexing.

galv commented:
I will finish reviewing this this evening. Thanks for the work.

tomkocse committed:
Add Init() to nnet-cudnn-simple-component with matrix_filename input; Add the test to diff the propagate output of cudnnConvolution and the original convolution component

tomkocse committed:
Fix the vectorization layout of the cudnn convolution component so now it has
same bitwise results as the original convolution component in both forward and backward operations.
Also the updated parameters are the same according to the same learning rate.
Acceptance of yzx vectorization is added. However, the notation in vectorization
should be changed as people usually arrange the letter of dimension from the largest to smallest
stride.

tomkocse committed:
Fixing the bug when the filter parameter matrix is calling its Read() to read the parameters from the file.
The returned matrix has no guarantee to be contiguous.

galv commented on src/nnet3/nnet-cudnn-simple-component.cc:
```diff
@@ -1,6 +1,7 @@
 // nnet3/nnet-simple-component.cc
```
nnet-**cudnn**-simple-component.cc

galv commented on src/nnet3/nnet-cudnn-simple-component.cc:
```diff
@@ -27,14 +28,61 @@
+  strides[4] = shape[1];
+}
+
+void array_strides_NXZYC(int n_dims, const int *shape, int *strides) {
```
This is currently unused. I haven't evaluated enough to understand why it's here.

tomkocse replied:
Sorry that i forget to remove this, but this will be useful if we allow user to select the output tensor layout.
Anyway, as you have suggested, these array_strides_xxxx() should be moved to another file if we are adding more cudnn routine (e.g. pooling).

galv commented on src/nnet3/nnet-test-utils.cc:
```diff
@@ -1323,11 +1326,11 @@ static void GenerateRandomCuDNNKaldiComponentPairConfig(std::string *component_t
          << " learning-rate=" << learning_rate;
 
-      os2 << " input-x-dim=" << input_x_dim
+      os2 << " input-x-dim=" << input_x_dim 
```
whitespace

galv commented on src/nnet3/nnet-cudnn-simple-component.cc:
```diff
@@ -294,6 +289,27 @@ void CuDNN3DConvolutionComponent::InitFromConfig(ConfigLine *cfl) {
   ok = ok && cfl->GetValue("filt-y-step", &filt_y_stride);
   ok = ok && cfl->GetValue("filt-z-step", &filt_z_stride);
 
+  // optional argument
```
This is optional because when you load the convolutional filter from a serialized matrix on disk, we expect only a single output channel, right? If so, please comment that.

tomkocse replied:
I change --input-num-filters into an optional field of the config line because i want to make the config line compatible with the original component. For the original component, the input-num-filters is always assumed to be one

galv replied:
Okay. Sounds good. It would be useful if you said that this was done specifically for compatibility with the existing component.
